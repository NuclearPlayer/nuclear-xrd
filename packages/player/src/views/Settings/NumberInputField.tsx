import { FC } from 'react';

import { Input } from '@nuclearplayer/ui';

type Props = {
  label: string;
  description?: string;
  value: number | string | undefined;
  setValue: (v: number) => void;
};

export const NumberInputField: FC<Props> = ({
  label,
  description,
  value,
  setValue,
}) => (
  <Input
    variant="number"
    label={label}
    description={description}
    value={String(value ?? '')}
    onChange={(e) => setValue(Number(e.target.value))}
  />
);
