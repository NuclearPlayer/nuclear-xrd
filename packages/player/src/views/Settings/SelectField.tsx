import { FC } from 'react';

import { Select } from '@nuclearplayer/ui';

type Option = { id: string; label: string };

type Props = {
  label: string;
  description?: string;
  value: string | undefined;
  setValue: (v: string) => void;
  options: Option[];
};

export const SelectField: FC<Props> = ({
  label,
  description,
  value,
  setValue,
  options,
}) => (
  <Select
    label={label}
    description={description}
    options={options}
    value={value ?? ''}
    onValueChange={setValue}
  />
);
