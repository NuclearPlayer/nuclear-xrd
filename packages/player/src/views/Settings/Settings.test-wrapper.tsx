import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import App from '../../App';
import { registerBuiltInCoreSettings } from '../../services/coreSettings';
import { initializeSettingsStore } from '../../stores/settingsStore';

export const SettingsWrapper = {
  async mount(): Promise<RenderResult> {
    await initializeSettingsStore();
    registerBuiltInCoreSettings();
    const component = render(<App />);
    await userEvent.click(
      await component.findByRole('button', { name: 'Preferences' }),
    );
    await userEvent.click(
      await component.findByRole('link', {
        name: 'Settings',
      }),
    );
    await screen.findByRole('heading', { name: 'Settings' });

    return component;
  },
};
