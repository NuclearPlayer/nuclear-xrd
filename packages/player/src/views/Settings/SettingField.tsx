import { FC } from 'react';

import type {
  EnumSettingDefinition,
  NumberWidget,
  SettingDefinition,
  SettingValue,
} from '@nuclearplayer/plugin-sdk';

import { NumberInputField } from './NumberInputField';
import { SelectField } from './SelectField';
import { SliderField } from './SliderField';
import { TextField } from './TextField';
import { ToggleField } from './ToggleField';
import { useSettingTranslation } from './useSettingTranslation';

type SettingFieldProps = {
  definition: SettingDefinition;
  value: SettingValue | undefined;
  setValue: (v: SettingValue) => void;
};

export const SettingField: FC<SettingFieldProps> = ({
  definition,
  value,
  setValue,
}) => {
  const { title: label, description } = useSettingTranslation(definition);
  const widgetType = definition.widget?.type;

  const renderers: Record<string, () => JSX.Element> = {
    toggle: () => (
      <ToggleField
        label={label}
        description={description}
        value={Boolean(value)}
        setValue={(v) => setValue(v)}
      />
    ),
    'number-input': () => (
      <NumberInputField
        label={label}
        description={description}
        value={value as number | string | undefined}
        setValue={(v) => setValue(v)}
      />
    ),
    slider: () => {
      const DEFAULT_MIN = 0;
      const DEFAULT_MAX = 100;
      const DEFAULT_STEP = 1;
      const widget = definition.widget as NumberWidget;
      const min = widget?.min ?? DEFAULT_MIN;
      const max = widget?.max ?? DEFAULT_MAX;
      const step = widget?.step ?? DEFAULT_STEP;
      const unit = widget?.unit;
      return (
        <SliderField
          label={label}
          value={Number(value ?? 0)}
          setValue={(v) => setValue(v)}
          min={min}
          max={max}
          step={step}
          unit={unit}
        />
      );
    },
    select: () => (
      <SelectField
        label={label}
        description={description}
        value={String(value ?? '')}
        setValue={(v) => setValue(v)}
        options={
          isEnumDefinition(definition)
            ? definition.options.map((o) => ({ id: o.value, label: o.label }))
            : []
        }
      />
    ),
    radio: () => (
      <SelectField
        label={label}
        description={description}
        value={String(value ?? '')}
        setValue={(v) => setValue(v)}
        options={
          isEnumDefinition(definition)
            ? definition.options.map((o) => ({ id: o.value, label: o.label }))
            : []
        }
      />
    ),
    password: () => (
      <TextField
        label={label}
        description={description}
        value={String(value ?? '')}
        setValue={(v) => setValue(v)}
        variant="password"
      />
    ),
    text: () => (
      <TextField
        label={label}
        description={description}
        value={String(value ?? '')}
        setValue={(v) => setValue(v)}
        variant="text"
      />
    ),
    textarea: () => (
      <TextField
        label={label}
        description={description}
        value={String(value ?? '')}
        setValue={(v) => setValue(v)}
        variant="text"
      />
    ),
  };

  const renderer = widgetType ? renderers[widgetType] : renderers.text;
  return renderer ? renderer() : null;
};

const isEnumDefinition = (
  def: SettingDefinition,
): def is EnumSettingDefinition => def.kind === 'enum';
