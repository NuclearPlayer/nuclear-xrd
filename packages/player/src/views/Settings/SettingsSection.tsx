import { FC } from 'react';

import { useSetting, type SettingDefinition } from '@nuclearplayer/plugin-sdk';
import { SectionShell } from '@nuclearplayer/ui';

import { coreSettingsHost } from '../../stores/settingsStore';
import { SettingField } from './SettingField';

type SettingsSectionProps = {
  title: string;
  settings: SettingDefinition[];
};

export const SettingsSection: FC<SettingsSectionProps> = ({
  title,
  settings,
}) => (
  <SectionShell title={title}>
    <div className="flex flex-col gap-6">
      {settings.map((definition) => {
        const [value, setValue] = useSetting(coreSettingsHost, definition.id);
        return (
          <SettingField
            key={definition.id}
            definition={definition}
            value={value}
            setValue={setValue}
          />
        );
      })}
    </div>
  </SectionShell>
);
