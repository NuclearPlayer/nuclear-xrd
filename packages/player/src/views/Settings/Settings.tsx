import { ScrollableArea, ViewShell } from '@nuclearplayer/ui';

import { SettingsSection } from './SettingsSection';
import { capitalize, useSettingsGroups } from './useSettingsGroups';

export const Settings = () => {
  const groups = useSettingsGroups();

  return (
    <ViewShell title="Settings">
      <ScrollableArea className="max-w-100 flex-1 overflow-hidden">
        {groups.map((group) => (
          <SettingsSection
            key={group.name}
            title={capitalize(group.name)}
            settings={group.settings}
          />
        ))}
      </ScrollableArea>
    </ViewShell>
  );
};
