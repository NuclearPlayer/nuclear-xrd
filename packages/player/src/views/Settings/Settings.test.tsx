import '../../test/setup';

import { vi } from 'vitest';

import { registerBuiltInCoreSettings } from '../../services/coreSettings';
import { resetInMemoryTauriStore } from '../../test/utils/inMemoryTauriStore';
import { SettingsWrapper } from './Settings.test-wrapper';

vi.mock('@tauri-apps/plugin-store', async () => {
  const mod = await import('../../test/utils/inMemoryTauriStore');
  return { LazyStore: mod.LazyStore };
});

window.scrollTo = vi.fn();

describe('Settings view', async () => {
  beforeEach(() => {
    resetInMemoryTauriStore();
    registerBuiltInCoreSettings();
  });
  it('(Snapshot) renders the settings view', async () => {
    const { asFragment } = await SettingsWrapper.mount();
    expect(asFragment()).toMatchSnapshot();
  });
});
