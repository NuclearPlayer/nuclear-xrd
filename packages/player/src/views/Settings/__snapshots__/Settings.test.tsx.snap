// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings view > (Snapshot) renders the settings view 1`] = `
<DocumentFragment>
  <div
    class="grid h-screen w-screen grid-rows-[auto_1fr_auto] overflow-hidden"
  >
    <header
      class="bg-background border-border flex h-12 items-center border-b-2 px-4"
    >
      <svg
        class="inline-flex h-6 w-6"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          stroke="currentColor"
          stroke-width="8"
          transform="matrix(.96157 0 0 .95989 1.9231 1.9198)"
        >
          <rect
            height="96"
            rx="19.2"
            ry="19.2"
            width="96"
            x="2"
            y="2"
          />
          <path
            d="m30 23.6 20 26.4 20-26.4zm20 26.4 20 26.4 20-26.4zm0 0h-40l20 26.4z"
          />
        </g>
      </svg>
      <div
        class="flex-1"
      />
      <button
        aria-checked="false"
        aria-label="Toggle theme"
        class="border-border relative inline-flex h-6 w-11 cursor-pointer items-center rounded-full border-2 transition-all focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-none data-checked:bg-primary bg-white data-disabled:cursor-not-allowed data-disabled:opacity-50 group"
        data-headlessui-state=""
        id="headlessui-switch-:r0:"
        role="switch"
        tabindex="0"
        type="button"
      >
        <span
          class="border-border pointer-events-none flex h-4 w-4 items-center justify-center rounded-full border-2 bg-white ring-0 transition-transform [&>svg]:fill-white dark:[&>svg]:fill-black translate-x-1 group-data-checked:translate-x-5"
        >
          <svg
            class="lucide lucide-sun text-foreground"
            fill="none"
            height="12"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="4"
            />
            <path
              d="M12 2v2"
            />
            <path
              d="M12 20v2"
            />
            <path
              d="m4.93 4.93 1.41 1.41"
            />
            <path
              d="m17.66 17.66 1.41 1.41"
            />
            <path
              d="M2 12h2"
            />
            <path
              d="M20 12h2"
            />
            <path
              d="m6.34 17.66-1.41 1.41"
            />
            <path
              d="m19.07 4.93-1.41 1.41"
            />
          </svg>
        </span>
      </button>
    </header>
    <div
      class="bg-background-secondary relative grid h-full min-h-0 grid-cols-[auto_1fr_auto]"
    >
      <div
        animate="[object Object]"
        class="bg-background-secondary border-border relative flex flex-col p-2 border-r-2"
        transition="[object Object]"
      >
        <span
          class="mb-4 flex flex-row items-center justify-end"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 right-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-left"
            type="submit"
          >
            <svg
              class="lucide lucide-panel-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M9 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          animate="[object Object]"
          class="flex-1 overflow-hidden"
          exit="[object Object]"
          initial="[object Object]"
          transition="[object Object]"
        >
          <div
            class="flex flex-1 flex-col items-start justify-start"
          >
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    class="lucide lucide-compass"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m16.24 7.76-1.804 5.411a2 2 0 0 1-1.265 1.265L7.76 16.24l1.804-5.411a2 2 0 0 1 1.265-1.265z"
                    />
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                  </svg>
                  Explore
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border border-2 h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    class="lucide lucide-settings"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
                    />
                    <circle
                      cx="12"
                      cy="12"
                      r="3"
                    />
                  </svg>
                  Preferences
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
              <div
                animate="[object Object]"
                class="border-l-foreground ml-4 flex flex-col gap-4 overflow-hidden border-l-2 pt-2"
                exit="[object Object]"
                initial="[object Object]"
                transition="[object Object]"
              >
                <a
                  aria-current="page"
                  class="active"
                  data-status="active"
                  href="/settings"
                >
                  <div
                    class="flex flex-row items-center justify-start gap-2 px-2"
                  >
                    <svg
                      class="lucide lucide-settings2"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M20 7h-9"
                      />
                      <path
                        d="M14 17H5"
                      />
                      <circle
                        cx="17"
                        cy="17"
                        r="3"
                      />
                      <circle
                        cx="7"
                        cy="7"
                        r="3"
                      />
                    </svg>
                    Settings
                  </div>
                </a>
                <a
                  href="/plugins"
                >
                  <div
                    class="flex flex-row items-center justify-start gap-2 px-2"
                  >
                    <svg
                      class="lucide lucide-blocks"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <rect
                        height="7"
                        rx="1"
                        width="7"
                        x="14"
                        y="3"
                      />
                      <path
                        d="M10 21V8a1 1 0 0 0-1-1H4a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-5a1 1 0 0 0-1-1H3"
                      />
                    </svg>
                    Plugins
                  </div>
                </a>
                <a
                  href="/themes"
                >
                  <div
                    class="flex flex-row items-center justify-start gap-2 px-2"
                  >
                    <svg
                      class="lucide lucide-palette"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="13.5"
                        cy="6.5"
                        fill="currentColor"
                        r=".5"
                      />
                      <circle
                        cx="17.5"
                        cy="10.5"
                        fill="currentColor"
                        r=".5"
                      />
                      <circle
                        cx="8.5"
                        cy="7.5"
                        fill="currentColor"
                        r=".5"
                      />
                      <circle
                        cx="6.5"
                        cy="12.5"
                        fill="currentColor"
                        r=".5"
                      />
                      <path
                        d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10c.926 0 1.648-.746 1.648-1.688 0-.437-.18-.835-.437-1.125-.29-.289-.438-.652-.438-1.125a1.64 1.64 0 0 1 1.668-1.668h1.996c3.051 0 5.555-2.503 5.555-5.554C21.965 6.012 17.461 2 12 2z"
                      />
                    </svg>
                    Themes
                  </div>
                </a>
              </div>
            </div>
            <div
              class="mb-4 flex w-full flex-col text-sm"
            >
              <button
                class="cursor-pointer rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent h-9 px-3 inline-flex flex-row items-center justify-between"
                data-headlessui-state=""
                type="submit"
              >
                <span
                  class="inline-flex flex-row items-center gap-2"
                >
                  <svg
                    class="lucide lucide-library"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m16 6 4 14"
                    />
                    <path
                      d="M12 6v14"
                    />
                    <path
                      d="M8 8v12"
                    />
                    <path
                      d="M4 4v16"
                    />
                  </svg>
                  Collection
                </span>
                <div
                  animate="[object Object]"
                  transition="[object Object]"
                >
                  <svg
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors right-0"
        />
      </div>
      <main
        class="bg-background-secondary overflow-auto"
        data-testid="player-workspace-main"
      >
        <div
          class="relative h-full w-full overflow-hidden"
        >
          <div
            animate="exit"
            class="absolute inset-0 h-full w-full"
            initial="enter"
            transition="[object Object]"
          >
            <div
              class="relative flex h-full w-full flex-col items-start justify-start px-6 pt-6"
            >
              <h1
                class="mb-6 flex w-full flex-0 flex-row text-center text-3xl font-bold"
              >
                Dashboard
              </h1>
              <div
                class="flex w-full flex-1 flex-col overflow-hidden"
              >
                Dashboard view
              </div>
            </div>
          </div>
          <div
            animate="center"
            class="absolute inset-0 h-full w-full"
            initial="enter"
            transition="[object Object]"
          >
            <div
              class="relative flex h-full w-full flex-col items-start justify-start px-6 pt-6"
            >
              <h1
                class="mb-6 flex w-full flex-0 flex-row text-center text-3xl font-bold"
              >
                Settings
              </h1>
              <div
                class="flex w-full flex-1 flex-col overflow-hidden"
              >
                <div
                  class="relative h-auto w-full max-w-100 flex-1 overflow-hidden"
                >
                  <div
                    class="scrollbar-hide flex h-full w-full flex-col overflow-auto"
                    style="scrollbar-width: none;"
                  >
                    <section
                      class="mb-6 flex w-full flex-col items-start justify-start"
                    >
                      <h2
                        class="mb-3 flex w-full flex-0 flex-row text-left text-2xl font-bold"
                      >
                        General
                      </h2>
                      <div
                        class="flex w-full flex-1 flex-col"
                      >
                        <div
                          class="flex flex-col gap-6"
                        >
                          <div
                            class="relative flex w-full flex-col gap-2"
                          >
                            <label
                              class="text-foreground text-sm font-semibold"
                              for="select-:r2:"
                              id="select-:r2:-label"
                            >
                              settings.general.language.title
                            </label>
                            <button
                              aria-expanded="false"
                              aria-haspopup="listbox"
                              class="border-border bg-primary text-foreground relative flex w-full items-center justify-between rounded border-2 text-left transition-colors focus-visible:ring-2 focus-visible:ring-black focus-visible:ring-offset-2 focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50 px-3 py-2 text-sm"
                              data-headlessui-state=""
                              id="select-:r2:"
                              type="button"
                            >
                              <span
                                class="block truncate"
                              >
                                English
                              </span>
                              <span
                                class="flex items-center"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="lucide lucide-chevron-down text-foreground transition-transform"
                                  fill="none"
                                  height="16"
                                  stroke="currentColor"
                                  stroke-linecap="round"
                                  stroke-linejoin="round"
                                  stroke-width="2"
                                  viewBox="0 0 24 24"
                                  width="16"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="m6 9 6 6 6-6"
                                  />
                                </svg>
                              </span>
                            </button>
                            <p
                              class="text-foreground-secondary text-sm select-none"
                              id="select-:r2:-description"
                            >
                              settings.general.language.description
                            </p>
                          </div>
                        </div>
                      </div>
                    </section>
                    <section
                      class="mb-6 flex w-full flex-col items-start justify-start"
                    >
                      <h2
                        class="mb-3 flex w-full flex-0 flex-row text-left text-2xl font-bold"
                      >
                        Playback
                      </h2>
                      <div
                        class="flex w-full flex-1 flex-col"
                      >
                        <div
                          class="flex flex-col gap-6"
                        >
                          <div
                            class="flex w-full flex-col gap-2"
                          >
                            <label
                              class="text-foreground text-sm font-semibold"
                              for="input-:r6:"
                              id="input-:r6:-label"
                            >
                              settings.playback.skipSeconds.title
                            </label>
                            <input
                              class="border-border bg-background-input text-foreground placeholder:text-foreground-secondary w-full rounded border-2 px-3 transition-colors focus-visible:ring-2 focus-visible:ring-black focus-visible:ring-offset-2 focus-visible:outline-none h-11"
                              data-headlessui-state=""
                              id="input-:r6:"
                              inputmode="numeric"
                              type="number"
                              value=""
                            />
                            <p
                              class="text-foreground-secondary text-sm select-none"
                              id="input-:r6:-description"
                            >
                              settings.playback.skipSeconds.description
                            </p>
                          </div>
                        </div>
                      </div>
                    </section>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
      <div
        animate="[object Object]"
        class="bg-background-secondary border-border relative flex flex-col p-2 border-l-2"
        transition="[object Object]"
      >
        <span
          class="mb-4 flex flex-row items-center justify-start"
        >
          <button
            class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center top-2 px-2 left-1"
            data-headlessui-state=""
            data-testid="sidebar-toggle-right"
            type="submit"
          >
            <svg
              class="lucide lucide-panel-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M15 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          animate="[object Object]"
          class="flex-1 overflow-hidden"
          exit="[object Object]"
          initial="[object Object]"
          transition="[object Object]"
        />
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors left-0"
        />
      </div>
    </div>
    <footer
      class="bg-background border-border flex h-16 items-center border-t-2 px-4"
    >
      <div
        class="grid w-full grid-cols-[minmax(0,1fr)_auto_minmax(0,1fr)] items-center gap-4"
      >
        <div
          class="min-w-0"
        >
          <div
            class="flex min-w-0 items-center gap-3"
          >
            <div
              class="border-border bg-background-secondary size-12 shrink-0 overflow-hidden rounded border-2"
            >
              <img
                alt=""
                class="size-full object-cover"
                src="https://picsum.photos/64"
              />
            </div>
            <div
              class="min-w-0"
            >
              <div
                class="text-foreground truncate text-sm font-bold"
              >
                Song Title
              </div>
              <div
                class="text-foreground-secondary truncate text-xs"
              >
                Artist Name
              </div>
            </div>
          </div>
        </div>
        <div
          class="justify-self-center"
        >
          <div
            class="flex items-center justify-center gap-2"
          >
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-shuffle"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 18h1.4c1.3 0 2.5-.6 3.3-1.7l6.1-8.6c.7-1.1 2-1.7 3.3-1.7H22"
                />
                <path
                  d="m18 2 4 4-4 4"
                />
                <path
                  d="M2 6h1.9c1.5 0 2.9.9 3.6 2.2"
                />
                <path
                  d="M22 18h-5.9c-1.3 0-2.6-.7-3.3-1.8l-.5-.8"
                />
                <path
                  d="m18 14 4 4-4 4"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-skip-back"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="19 20 9 12 19 4 19 20"
                />
                <line
                  x1="5"
                  x2="5"
                  y1="19"
                  y2="5"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-play"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="6 3 20 12 6 21 6 3"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-skip-forward"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="5 4 15 12 5 20 5 4"
                />
                <line
                  x1="19"
                  x2="19"
                  y1="5"
                  y2="19"
                />
              </svg>
            </button>
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-repeat"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m17 2 4 4-4 4"
                />
                <path
                  d="M3 11v-1a4 4 0 0 1 4-4h14"
                />
                <path
                  d="m7 22-4-4 4-4"
                />
                <path
                  d="M21 13v1a4 4 0 0 1-4 4H3"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="justify-self-end"
        >
          <div
            class="flex items-center gap-2"
          >
            <button
              class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-transparent size-10 justify-center"
              data-headlessui-state=""
              type="submit"
            >
              <svg
                class="lucide lucide-volume2"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
                />
                <path
                  d="M16 9a5 5 0 0 1 0 6"
                />
                <path
                  d="M19.364 18.364a9 9 0 0 0 0-12.728"
                />
              </svg>
            </button>
            <div
              class="w-24"
            >
              <div
                class="flex w-full flex-col gap-2"
              >
                <div
                  class="relative h-6 w-full overscroll-contain"
                >
                  <div
                    class="pointer-events-none absolute inset-x-0 top-1/2 -translate-y-1/2"
                  >
                    <div
                      class="border-border bg-background-input relative h-3 w-full rounded border-2"
                    >
                      <div
                        class="bg-primary absolute top-0 left-0 h-full rounded-l"
                        style="width: 75%;"
                      />
                      <div
                        class="border-border absolute top-1/2 -ml-2 h-5 w-5 -translate-y-1/2 rounded-full border-2 bg-white"
                        style="left: 75%;"
                      />
                    </div>
                  </div>
                  <input
                    aria-labelledby=":r1:-label"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="75"
                    aria-valuetext="75"
                    class="absolute inset-0 z-[1] h-6 w-full appearance-none opacity-0 cursor-pointer"
                    id=":r1:"
                    max="100"
                    min="0"
                    step="1"
                    type="range"
                    value="75"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </footer>
    <section
      aria-atomic="false"
      aria-label="Notifications alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;
